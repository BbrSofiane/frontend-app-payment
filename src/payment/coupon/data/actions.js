import { AsyncActionType } from '../../../common/utils';

export const ADD_COUPON = new AsyncActionType('PAYMENT', 'ADD_COUPON');
export const REMOVE_COUPON = new AsyncActionType('PAYMENT', 'REMOVE_COUPON');
export const UPDATE_COUPON_DRAFT = 'PAYMENT__UPDATE_COUPON_DRAFT';

// ADD COUPON

export const addCoupon = code => ({
  type: ADD_COUPON.BASE,
  payload: {
    code,
  },
});

export const addCouponBegin = () => ({
  type: ADD_COUPON.BEGIN,
});

export const addCouponSuccess = (id, code, benefitValue) => ({
  type: ADD_COUPON.SUCCESS,
  payload: {
    benefitValue,
    code,
    id,
  },
});

export const addCouponFailure = errorCode => ({
  type: ADD_COUPON.FAILURE,
  payload: {
    errorCode,
  },
});

// REMOVE COUPON

export const removeCoupon = id => ({
  type: REMOVE_COUPON.BASE,
  payload: {
    id,
  },
});

export const removeCouponBegin = () => ({
  type: REMOVE_COUPON.BEGIN,
});

export const removeCouponSuccess = result => ({
  type: REMOVE_COUPON.SUCCESS,
  payload: {
    result,
  },
});

export const removeCouponFailure = errorCode => ({
  type: REMOVE_COUPON.FAILURE,
  payload: {
    errorCode,
  },
});

// UPDATE COUPON DRAFT

export const updateCouponDraft = code => ({
  type: UPDATE_COUPON_DRAFT,
  payload: {
    code,
  },
});
