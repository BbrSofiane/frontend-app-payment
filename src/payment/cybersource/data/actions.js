import { utils } from '../../../common';

const { AsyncActionType } = utils;

export const SUBMIT_PAYMENT_CYBERSOURCE = new AsyncActionType('PAYMENT', 'SUBMIT_PAYMENT_CYBERSOURCE');

export const submitPaymentCybersource = (cardHolderInfo, cardDetails) => ({
  type: SUBMIT_PAYMENT_CYBERSOURCE.BASE,
  payload: {
    cardHolderInfo,
    cardDetails,
  },
});

export const submitPaymentCybersourceBegin = () => ({
  type: SUBMIT_PAYMENT_CYBERSOURCE.BEGIN,
});

export const submitPaymentCybersourceFailure = () => ({
  type: SUBMIT_PAYMENT_CYBERSOURCE.FAILURE,
});
