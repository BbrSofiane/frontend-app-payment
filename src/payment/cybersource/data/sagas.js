import { call, put, select, takeEvery } from 'redux-saga/effects';
import { logError } from '@edx/frontend-logging';

import {
  SUBMIT_PAYMENT_CYBERSOURCE,
  submitPaymentCybersourceBegin,
  submitPaymentCybersourceFailure,
} from './actions';

import * as CybersourceApiService from './service';

import { generateAndSubmitForm } from '../../../common/utils';
import { handleErrorFeedback } from '../../../feedback';
import { configuration } from '../../../environment';
import { isUnknownError } from '../../../common/serviceUtils';

export function* handleSubmitPaymentCybersource(action) {
  const { cardHolderInfo, cardDetails } = action.payload;
  yield put(submitPaymentCybersourceBegin());
  try {
    const sdnCheck = yield call(
      CybersourceApiService.sdnCheck,
      cardHolderInfo.firstName,
      cardHolderInfo.lastName,
      cardHolderInfo.city,
      cardHolderInfo.country,
    );

    if (sdnCheck.hits > 0) {
      window.location.href = `${configuration.ECOMMERCE_BASE_URL}/payment/sdn/failure/`;
    }

    const basket = yield select(state => ({ ...state.payment.basket }));
    const checkout = yield call(
      CybersourceApiService.checkout,
      basket.basketId,
      cardHolderInfo,
    );

    const {
      cardNumber,
      cardTypeId,
      securityCode,
      cardExpirationMonth,
      cardExpirationYear,
    } = cardDetails;
    const cybersourcePaymentParams = {
      ...checkout.form_fields,
      card_number: cardNumber,
      card_type: cardTypeId,
      card_cvn: securityCode,
      card_expiry_date: [cardExpirationMonth.padStart(2, '0'), cardExpirationYear].join('-'),
    };
    generateAndSubmitForm(configuration.CYBERSOURCE_URL, cybersourcePaymentParams);
  } catch (error) {
    yield put(submitPaymentCybersourceFailure());
    yield call(handleErrorFeedback, error, true);
    if (isUnknownError(error)) {
      logError(error);
    }
  }
}

export default function* saga() {
  yield takeEvery(SUBMIT_PAYMENT_CYBERSOURCE.BASE, handleSubmitPaymentCybersource);
}
