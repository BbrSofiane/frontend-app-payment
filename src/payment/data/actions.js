import { utils } from '../../common';

const { AsyncActionType } = utils;

export const FETCH_BASKET = new AsyncActionType('PAYMENT', 'FETCH_BASKET');
export const SUBMIT_PAYMENT = new AsyncActionType('PAYMENT', 'SUBMIT_PAYMENT');
export const CHECKOUT = new AsyncActionType('PAYMENT', 'CHECKOUT');


export const fetchBasket = () => ({
  type: FETCH_BASKET.BASE,
  payload: {},
});

export const fetchBasketBegin = () => ({
  type: FETCH_BASKET.BEGIN,
});

export const fetchBasketSuccess = result => ({
  type: FETCH_BASKET.SUCCESS,
  payload: result,
});

export const fetchBasketFailure = () => ({
  type: FETCH_BASKET.FAILURE,
});

export const submitPayment = cardHolderInfo => ({
  type: SUBMIT_PAYMENT.BASE,
  payload: {
    cardHolderInfo,
  },
});

export const submitPaymentBegin = () => ({
  type: SUBMIT_PAYMENT.BEGIN,
});

export const submitPaymentFailure = () => ({
  type: SUBMIT_PAYMENT.FAILURE,
});

export const checkoutSuccess = (paymentProcessorUrl, paymentProcessorFormFields) => ({
  type: CHECKOUT.SUCCESS,
  payload: {
    paymentProcessorUrl,
    paymentProcessorFormFields,
  },
});
