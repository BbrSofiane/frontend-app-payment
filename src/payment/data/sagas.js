import { all, call, put, takeEvery } from 'redux-saga/effects';

// Actions
import {
  FETCH_BASKET,
  fetchBasketBegin,
  fetchBasketSuccess,
  fetchBasketFailure,
} from './actions';

// Services
import * as PaymentApiService from './service';

import { saga as couponSaga, addCouponSuccess, addCouponBegin } from '../coupon';
import { saga as cybersourceSaga } from '../cybersource';
import { saga as payPalSaga } from '../paypal';
import { handleErrors, handleMessages } from '../../feedback';

export function* handleFetchBasket() {
  yield put(fetchBasketBegin());
  yield put(addCouponBegin());
  try {
    const result = yield call(PaymentApiService.getBasket);
    yield put(fetchBasketSuccess(result));
    yield call(handleMessages, result.messages);
    if (result.coupons.length === 0) {
      yield put(addCouponSuccess(null, null, null));
    } else {
      yield put(addCouponSuccess(
        result.coupons[0].id,
        result.coupons[0].code,
        result.coupons[0].benefitValue,
      ));
    }
  } catch (e) {
    yield put(fetchBasketFailure());
    yield call(handleErrors, e);
  }
}

export default function* saga() {
  yield takeEvery(FETCH_BASKET.BASE, handleFetchBasket);
  yield all([
    couponSaga(),
    cybersourceSaga(),
    payPalSaga(),
  ]);
}
