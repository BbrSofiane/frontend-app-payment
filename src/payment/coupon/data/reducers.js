import { ADD_COUPON, REMOVE_COUPON, UPDATE_COUPON_DRAFT } from './actions';

const defaultState = {
  benefitValue: null,
  code: null,
  id: null,
  errorCode: null,
  loaded: false,
  loading: false,
};

const reducer = (state = defaultState, action = null) => {
  if (action !== null) {
    switch (action.type) {
      case ADD_COUPON.BEGIN:
        return {
          ...state,
          loading: true,
          loaded: false,
          errorCode: null,
        };
      case ADD_COUPON.SUCCESS:
        return {
          ...state,
          loading: false,
          loaded: true,
          errorCode: null,
          benefitValue: action.payload.benefitValue,
          code: action.payload.code,
          id: action.payload.id,
        };
      case ADD_COUPON.FAILURE:
        return {
          ...state,
          loading: false,
          loaded: false,
          errorCode: action.payload.errorCode,
          id: null,
        };
      case REMOVE_COUPON.BEGIN:
        return {
          ...state,
          loading: true,
          loaded: false,
          errorCode: null,
        };
      case REMOVE_COUPON.SUCCESS:
        return defaultState;
      case REMOVE_COUPON.FAILURE:
        return {
          ...state,
          loading: false,
          loaded: true, // There's still a loaded coupon.
          errorCode: action.payload.errorCode,
        };
      case UPDATE_COUPON_DRAFT:
        return {
          ...state,
          code: action.payload.code,
        };
      default:
    }
  }

  return state;
};

export default reducer;
