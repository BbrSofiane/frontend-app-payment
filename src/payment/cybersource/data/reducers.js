import { SUBMIT_PAYMENT_CYBERSOURCE } from './actions';

const defaultState = {
  submitting: false,
};

const reducer = (state = defaultState, action = null) => {
  if (action !== null) {
    switch (action.type) {
      case SUBMIT_PAYMENT_CYBERSOURCE.BEGIN:
        return {
          ...state,
          submitting: true,
        };
      case SUBMIT_PAYMENT_CYBERSOURCE.FAILURE:
        return {
          ...state,
          submitting: false,
        };
      default:
    }
  }

  return state;
};

export default reducer;
