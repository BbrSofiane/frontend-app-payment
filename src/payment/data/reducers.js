import { combineReducers } from 'redux';

import { FETCH_BASKET } from './actions';
import { reducer as coupon } from '../coupon';
import { reducer as cybersource } from '../cybersource';
import { reducer as paypal } from '../paypal';

export const basketInitialState = {
  loading: false,
  submitting: false,
  products: [],
};

const basket = (state = basketInitialState, action = null) => {
  switch (action.type) {
    case FETCH_BASKET.BEGIN:
      return {
        ...state,
        loading: true,
        loaded: false,
      };
    case FETCH_BASKET.SUCCESS:
      return {
        ...state,
        ...action.payload,
        loading: false,
        loaded: true,
      };
    case FETCH_BASKET.FAILURE:
      return {
        ...state,
        loading: false,
        loaded: false,
      };
    default:
      return state;
  }
};

const reducer = combineReducers({
  basket,
  coupon,
  cybersource,
  paypal,
});

export default reducer;
