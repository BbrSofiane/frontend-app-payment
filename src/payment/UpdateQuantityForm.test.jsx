/* eslint-disable global-require */
import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';

import UpdateQuantityForm from './UpdateQuantityForm';
import * as PaymentApiService from './data/service';

const mockStore = configureMockStore();
const store = mockStore(require('./__mocks__/defaultState.mockStore.js'));

PaymentApiService.postQuantity = jest.fn();

describe('<UpdateQuantityForm/>', () => {
  it('should render properly', () => {
    const tree = renderer
      .create((
        <Provider store={store} >
          <UpdateQuantityForm
            summaryQuantity={1}
          />
        </Provider >
      ))
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should send an update quantity request on click', () => {
    const mockApiClient = {
      post: jest.fn(),
    };

    const newQuantity = 1234;

    const wrapper = mount((
      <Provider store={store} >
        <UpdateQuantityForm
          summaryQuantity={newQuantity}
        />
      </Provider >
    ));
    wrapper.find('button').simulate('click');
    expect(PaymentApiService.postQuantity).toHaveBeenCalledWith(124555);
  });
});
