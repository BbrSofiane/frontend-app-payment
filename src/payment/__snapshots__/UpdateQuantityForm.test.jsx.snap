// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpdateQuantityForm/> should render properly 1`] = `
<form
  className="summary-row form-inline"
  onSubmit={[Function]}
>
  <div
    className="form-group mr-2"
  >
    <label
      htmlFor="code-quantity"
    >
      <span>
        Quantity
      </span>
    </label>
    <div
      className="position-relative"
    >
      <input
        className="form-control form-control-sm"
        defaultValue={1}
        id="code-quantity"
        max="100"
        min="1"
        name="code-quantity"
        style={
          Object {
            "width": "5rem",
          }
        }
        type="number"
      />
      <small
        className="text-muted small"
        style={
          Object {
            "left": 0,
            "position": "absolute",
            "top": "100%",
          }
        }
      >
        <span>
          Max: 100
        </span>
      </small>
    </div>
  </div>
  <button
    aria-disabled={false}
    aria-live="assertive"
    className="btn pgn__stateful-btn pgn__stateful-btn-state-default btn-primary btn-sm"
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="submit"
  >
    <span
      className="d-flex align-items-center justify-content-center"
    >
      <span>
        Update
      </span>
    </span>
  </button>
</form>
`;
